File: front/src/mock/data/stocks.ts
import type { Stock } from '@/api/types'

/** A 股真实代码 + 行业 — mock 基础数据 */
export const stocks: Stock[] = [
  { code: 'sz.000001', name: '平安银行', industries: ['银行'], price: 11.85, change_pct: 1.23 },
  { code: 'sh.600519', name: '贵州茅台', industries: ['白酒', '食品饮料'], price: 1685.50, change_pct: -0.45 },
  { code: 'sz.000858', name: '五粮液', industries: ['白酒', '食品饮料'], price: 142.30, change_pct: 0.89 },
  { code: 'sh.600036', name: '招商银行', industries: ['银行'], price: 35.67, change_pct: 0.56 },
  { code: 'sz.000333', name: '美的集团', industries: ['家电'], price: 65.42, change_pct: -1.20 },
  { code: 'sh.600276', name: '恒瑞医药', industries: ['医药'], price: 48.15, change_pct: 2.34 },
  { code: 'sz.000725', name: '京东方A', industries: ['电子', '半导体'], price: 4.32, change_pct: 0.88 },
  { code: 'sh.601318', name: '中国平安', industries: ['保险', '金融'], price: 48.90, change_pct: -0.67 },
  { code: 'sz.002594', name: '比亚迪', industries: ['汽车', '新能源'], price: 245.60, change_pct: 3.45 },
  { code: 'sh.600887', name: '伊利股份', industries: ['食品饮料'], price: 28.75, change_pct: 0.12 },
  { code: 'sz.000568', name: '泸州老窖', industries: ['白酒'], price: 178.20, change_pct: 1.56 },
  { code: 'sh.601166', name: '兴业银行', industries: ['银行'], price: 18.43, change_pct: -0.34 },
  { code: 'sz.002415', name: '海康威视', industries: ['电子', '安防'], price: 32.18, change_pct: 0.78 },
  { code: 'sh.600030', name: '中信证券', industries: ['证券'], price: 25.34, change_pct: 1.89 },
  { code: 'sz.000651', name: '格力电器', industries: ['家电'], price: 38.92, change_pct: -0.56 },
  { code: 'sh.600009', name: '上海机场', industries: ['交通运输'], price: 52.10, change_pct: 0.45 },
  { code: 'sz.002230', name: '科大讯飞', industries: ['计算机', '人工智能'], price: 45.67, change_pct: 2.78 },
  { code: 'sh.600585', name: '海螺水泥', industries: ['建材'], price: 28.45, change_pct: -1.10 },
  { code: 'sz.300750', name: '宁德时代', industries: ['新能源', '电池'], price: 185.30, change_pct: 1.67 },
  { code: 'sh.601012', name: '隆基绿能', industries: ['新能源', '光伏'], price: 22.34, change_pct: -2.30 },
  { code: 'sz.000625', name: '长安汽车', industries: ['汽车'], price: 13.56, change_pct: 0.92 },
  { code: 'sh.600690', name: '海尔智家', industries: ['家电'], price: 25.78, change_pct: 0.34 },
  { code: 'sz.002241', name: '歌尔股份', industries: ['电子', '消费电子'], price: 23.45, change_pct: 1.23 },
  { code: 'sh.600809', name: '山西汾酒', industries: ['白酒'], price: 218.50, change_pct: 0.89 },
  { code: 'sz.000538', name: '云南白药', industries: ['医药'], price: 52.30, change_pct: -0.45 },
  { code: 'sh.601888', name: '中国中免', industries: ['商业百货'], price: 85.60, change_pct: 1.34 },
  { code: 'sz.002714', name: '牧原股份', industries: ['农业', '养殖'], price: 45.20, change_pct: -1.56 },
  { code: 'sh.600436', name: '片仔癀', industries: ['医药'], price: 235.40, change_pct: 0.67 },
  { code: 'sz.300059', name: '东方财富', industries: ['证券', '互联网金融'], price: 14.56, change_pct: 2.10 },
  { code: 'sh.600900', name: '长江电力', industries: ['电力'], price: 25.40, change_pct: 0.23 },
]

/** 按代码查找 */
export function findStock(code: string): Stock | undefined {
  return stocks.find((s) => s.code === code)
}

/** 所有行业列表 */
export const allIndustries: string[] = Array.from(
  new Set(stocks.flatMap((s) => s.industries)),
).sort()

File: WebAPI/routers/stocks.py
# -*- coding: utf-8 -*-
"""
Stock 路由 — 股票管理 API。

注意: 静态路由（/status, /export-config）必须在参数化路由（/{code}）之前定义。
"""

from typing import Optional

from fastapi import APIRouter, HTTPException, Query

from ..stock_store import (
    delete_stock,
    get_stock,
    get_stocks_with_daily_close,
    import_stocks,
    list_stocks,
    upsert_industries,
    upsert_stock,
)

router = APIRouter(prefix="/api/stocks", tags=["stocks"])


# ---- 静态路由（必须在参数化路由之前）----


@router.get("/industries")
async def stock_industries():
    """行业列表（完整实现在 chan-stock-manage）。"""
    return []


@router.get("/status")
async def stocks_status():
    """灌数状态（返回 DuckDB 统计信息）。"""
    from ..config import DUCKDB_PATH
    import os

    if not os.path.exists(DUCKDB_PATH):
        return {"total_stocks": 0, "total_bars": 0, "last_update": None}

    try:
        import duckdb

        conn = duckdb.connect(DUCKDB_PATH, read_only=True)
        try:
            row = conn.execute(
                "SELECT COUNT(DISTINCT code), COUNT(*), MAX(time_key) FROM kline"
            ).fetchone()
            return {
                "total_stocks": row[0] if row else 0,
                "total_bars": row[1] if row else 0,
                "last_update": row[2] if row else None,
            }
        finally:
            conn.close()
    except Exception:
        return {"total_stocks": 0, "total_bars": 0, "last_update": None}


@router.get("/export-config")
async def stocks_export_config():
    """导出灌数配置。"""
    from ..stock_store import _get_pg_conn

    pg_conn = _get_pg_conn()
    if pg_conn is not None:
        try:
            with pg_conn.cursor() as cur:
                cur.execute("SELECT code, kl_types, autypes FROM stock WHERE enabled = TRUE")
                rows = cur.fetchall()
        finally:
            pg_conn.close()
        return [
            {
                "code": r[0],
                "kl_types": r[1] or ["K_DAY"],
                "autypes": r[2] or ["QFQ"],
            }
            for r in rows
        ]

    from ..config import DUCKDB_PATH
    import os

    if os.path.exists(DUCKDB_PATH):
        try:
            import duckdb

            conn = duckdb.connect(DUCKDB_PATH, read_only=True)
            try:
                rows = conn.execute(
                    "SELECT DISTINCT code FROM kline ORDER BY code"
                ).fetchall()
                return [
                    {"code": r[0], "kl_types": ["K_DAY"], "autypes": ["QFQ"]}
                    for r in rows
                ]
            finally:
                conn.close()
        except Exception:
            pass
    return []


@router.post("/import")
async def import_stocks_endpoint(body: dict):
    """批量导入股票。"""
    stocks = body.get("stocks", [])
    return import_stocks(stocks)


# ---- 集合路由 ----


@router.get("")
async def list_stocks_endpoint(
    q: str = Query(""),
    exchange: str = Query(""),
    industry: str = Query(""),
    enabled: Optional[bool] = Query(None),
    page: int = Query(1),
    page_size: int = Query(20),
):
    """股票列表（筛选 + 分页）。"""
    return list_stocks(
        q=q,
        exchange=exchange,
        industry=industry,
        enabled=enabled,
        page=page,
        page_size=page_size,
    )


@router.post("")
async def upsert_stock_endpoint(body: dict):
    """新增或更新股票。"""
    code = body.get("code", "")
    if not code:
        raise HTTPException(status_code=422, detail="code is required")
    stock = upsert_stock(
        code=code,
        name=body.get("name", code),
        exchange=body.get("exchange", ""),
        enabled=body.get("enabled", True),
        kl_types=body.get("kl_types"),
        autypes=body.get("autypes"),
        tags=body.get("tags"),
        notes=body.get("notes", ""),
    )
    industries_data = body.get("industries", [])
    if industries_data:
        upsert_industries(code, industries_data)
    return stock


# ---- 参数化路由（/{code} 必须在最后）----


@router.get("/{code}/profile")
async def stock_profile(code: str):
    """标的详情（供前端 K 线页 profile）。"""
    stock = get_stock(code)
    if not stock:
        return {
            "code": code,
            "name": code,
            "industries": [],
            "region": "--",
            "concepts": [],
            "price": 0,
            "change_pct": 0,
        }

    profiles = get_stocks_with_daily_close([code])
    if profiles:
        return profiles[0]
    return stock


@router.get("/{code}")
async def get_stock_endpoint(code: str):
    """股票详情。"""
    stock = get_stock(code)
    if not stock:
        raise HTTPException(status_code=404, detail=f"Stock {code!r} not found")
    return stock


@router.delete("/{code}")
async def delete_stock_endpoint(code: str):
    """删除股票。"""
    ok = delete_stock(code)
    if not ok:
        raise HTTPException(status_code=404, detail=f"Stock {code!r} not found")
    return {"success": True}